src/aind_ephys_ibl_gui_conversion/ephys.py

```diff
@@ -2,36 +2,76 @@
 Functions to process ephys data
 """
 
+import json
 import logging
 import re
+import shutil
+from collections import defaultdict
+from datetime import datetime
 from pathlib import Path
-from typing import Union
+from typing import List, Union
 
 import numpy as np
 import pandas as pd
 import spikeinterface as si
 import spikeinterface.extractors as se
 import spikeinterface.preprocessing as spre
+from aind_data_schema.core.data_description import DerivedDataDescription
+from aind_data_schema.core.processing import (
+    DataProcess,
+    PipelineProcess,
+    Processing,
+)
+from aind_data_schema_models.process_names import ProcessName
+from aind_metadata_upgrader.data_description_upgrade import (
+    DataDescriptionUpgrade,
+)
 from scipy.signal import welch
 from spikeinterface.core import get_random_data_chunks
 from spikeinterface.exporters import export_to_phy
 from spikeinterface.exporters.to_ibl import compute_rms
 
-# here we define some constants used for defining if timestamps are ok
-# or should be skipped
-ACCEPTED_NEGATIVE_DEVIATION_MS = (
-    0.2  # we allow for small negative timestamps diff glitches
-)
-# maximum number of negative timestamps allowed below the accepted deviation
-MAX_NUM_NEGATIVE_TIMESTAMPS = 10
-ABS_MAX_TIMESTAMPS_DEVIATION_MS = (
-    2  # absolute maximum deviation allowed for timestamps (also positive)
-)
+STREAM_PROBE_REGEX = re.compile(r"^Record Node \d+#[^.]+\.(.+?)(-AP|-LFP)?$")
 
-MAX_NUM_NEGATIVE_TIMESTAMPS = 10
-MAX_TIMESTAMPS_DEVIATION_MS = 1
 
-STREAM_PROBE_REGEX = re.compile(r"^Record Node \d+#[^.]+\.(.+?)(-AP|-LFP)?$")
+def _merge_dicts(d1: defaultdict, d2: defaultdict) -> defaultdict:
+    """
+    Merge two ``defaultdict(list)`` objects into a new ``defaultdict(list)``.
+
+    This function returns a new ``defaultdict(list)`` containing all key–value
+    pairs from both input dictionaries. If a key appears in both, their lists
+    are concatenated in order (values from ``d1`` first, then ``d2``). The
+    inputs are not modified.
+
+    Used for keeping track of mappings when surface recroding is a seperate
+    data asset.
+
+    Parameters
+    ----------
+    d1 : defaultdict
+        The first ``defaultdict(list)`` to merge.
+    d2 : defaultdict
+        The second ``defaultdict(list)`` to merge.
+
+    Returns
+    -------
+    defaultdict
+        A new ``defaultdict(list)`` containing the merged key–value pairs.
+
+    Examples
+    --------
+    >>> from collections import defaultdict
+    >>> a = defaultdict(list, {'x': [1, 2], 'y': [3]})
+    >>> b = defaultdict(list, {'y': [4], 'z': [5]})
+    >>> merge_defaultdicts(a, b)
+    defaultdict(<class 'list'>, {'x': [1, 2], 'y': [3, 4], 'z': [5]})
+    """
+    merged = defaultdict(list)
+    for k, v in d1.items():
+        merged[k].extend(v)
+    for k, v in d2.items():
+        merged[k].extend(v)
+    return merged
```

Comment:
This function is likely unnecessary. It reimplements the built in union operator `d1 | d2`, which works with `defaultdict`. The subtlety here with `defaultdict` is that the `default_factory` of the first argument to `|` is used, while this function hard-codes `list` as the `default_factory`. I didn't check the call sites, but if they also have  `default_factory=list` then I think you should remove this function.

```suggestion
```